### ISSUE TITLE
🐛 Bug Report — Runtime APIs - WritableStream DefaultWriter await write() does not copy chunk buffer

### ISSUE THREAD

**Issue by pmeenan** [cloudflare/workerd#3137]
If you write a buffer to a TransformStream's WritableStream with `await writer.write()` and then modify the buffer after the promise completes, the output stream will be corrupted.

As best as I can tell [from the spec](https://streams.spec.whatwg.org/#default-writer-prototype), it should be safe to modify the chunk after the promise resolves:

>If chunk is mutable, [producers](https://streams.spec.whatwg.org/#producer) are advised to avoid mutating it after passing it to [write()](https://streams.spec.whatwg.org/#default-writer-write), until after the promise returned by [write()](https://streams.spec.whatwg.org/#default-writer-write) settles. This ensures that the [underlying sink](https://streams.spec.whatwg.org/#underlying-sink) receives and processes the same value that was passed in.

As it stands currently, the only safe way to do it is to `slice()` the buffer to make a copy of it when you pass it to write().

This was discovered as part of working on a [worker that uses wasm to compress the response streams](https://github.com/pmeenan/dictionary-worker/blob/main/src/index.js#L285) in a passthrough transform stream. The resulting stream would be corrupted with chunks being overwritten with later data.

**Comment by jasnell** [comment]
Yes, this is a known limitation of the existing implementation that we end up having to preserve for backwards compatibility :-/ ... The other approach you could take aside from `slice()` is to copy the input like `writer.write(new Uint8Array(chunk))`. We've considered adding a fix for this behind a compatibility flag but it wasn't clear if folks were actually bumping up against this or not. I think at this point it's clear we should.

**Comment by pmeenan** [comment]
Just documenting the behavior difference on the docs would probably be enough if you need to maintain it for compatibility reasons. It's just a pain to debug when it happens.
